mono/metadata/android.c

```diff
@@ -0,0 +1,537 @@
+# else
+	return CPU_KIND_UNKNOWN;
+# endif
+#elif HOST_DARWIN
+# if __x86_64__
+	return CPU_KIND_X86_64;
+# elif __i386__
+	return CPU_KIND_X86;
+# else
+	return CPU_KIND_UNKNOWN;
+# endif
+#elif HOST_ANDROID
+# if __arm__
+	return CPU_KIND_ARM;
+# elif __aarch64__
+	return CPU_KIND_ARM64;
+# elif __x86_64__
+	return CPU_KIND_X86_64;
+# elif __i386__
+	return CPU_KIND_X86;
+# elif __mips__
+	return CPU_KIND_MIPS;
+# else
+	return CPU_KIND_UNKNOWN;
+# endif
+#else
+# error get_built_for_cpu not implemented
+#endif // HOST_WIN32
+}
+
+static guint16
+get_running_on_cpu (void)
+{
+#ifdef HOST_WIN32
+	SYSTEM_INFO si;
+
+	GetSystemInfo (&si);
+	switch (si.wProcessorArchitecture) {
+		case PROCESSOR_ARCHITECTURE_AMD64:
+			return CPU_KIND_X86_64;
+		case PROCESSOR_ARCHITECTURE_ARM:
+			return CPU_KIND_ARM;
+		case PROCESSOR_ARCHITECTURE_INTEL:
+			return CPU_KIND_X86;
+		default:
+			return CPU_KIND_UNKNOWN;
+	}
+#elif HOST_DARWIN
+	cpu_type_t cputype;
+	size_t length;
+
+	length = sizeof (cputype);
+	sysctlbyname ("hw.cputype", &cputype, &length, NULL, 0);
+	switch (cputype) {
+		case CPU_TYPE_X86:
+			return CPU_KIND_X86;
+		case CPU_TYPE_X86_64:
+			return CPU_KIND_X86_64;
+		default:
+			return CPU_KIND_UNKNOWN;
+	}
+#elif HOST_ANDROID
+# if __arm__
+	if (!detect_houdini ()) {
+		return CPU_KIND_ARM;
+	} else {
+		/* If houdini is mapped in we're running on x86 */
+		return CPU_KIND_X86;
+	}
+# elif __aarch64__
+	return CPU_KIND_ARM64;
+# elif __x86_64__
+	return CPU_KIND_X86_64;
+# elif __i386__
+	return is_64_bit () ? CPU_KIND_X86_64 : CPU_KIND_X86;
+# elif __mips__
+	return CPU_KIND_MIPS;
+# endif
+#else
+# error get_running_on_cpu not implemented
+#endif // HOST_WIN32
+}
+
+void
+_monodroid_detect_cpu_and_architecture (guint16 *built_for_cpu, guint16 *running_on_cpu, guint8 *is64bit)
+{
+	g_assert (is64bit);
+	*is64bit = (guint8) is_64_bit ();
+	g_assert (built_for_cpu);
+	*built_for_cpu = get_built_for_cpu ();
+	g_assert (running_on_cpu);
+	*running_on_cpu = get_running_on_cpu ();
+}
+
+gint32
+_monodroid_get_dns_servers (gpointer *dns_servers_array)
+{
+	g_assert (dns_servers_array);
+	*dns_servers_array = NULL;
+
+	gsize  len;
+	gchar *dns;
+	gchar *dns_servers [8];
+	gint   count = 0;
+	gchar  prop_name[] = "net.dnsX";
+	for (gint i = 0; i < 8; i++) {
+		prop_name [7] = (char)(i + 0x31);
+		len = monodroid_get_system_property (prop_name, &dns);
+		if (len <= 0) {
+			dns_servers [i] = NULL;
+			continue;
+		}
+		dns_servers [i] = g_strndup (dns, len);
+		count++;
+	}
+
+	if (count <= 0)
+		return 0;
+
+	gchar **ret = g_new (gchar*, count);
+	gchar **p = ret;
+	for (gint i = 0; i < 8; i++) {
+		if (!dns_servers [i])
+			continue;
+		*p++ = dns_servers [i];
+	}
+
+	*dns_servers_array = (gpointer)ret;
+	return count;
+}
+
+static MonoBoolean
+_monodroid_get_network_interface_state (const gchar *ifname, MonoBoolean *is_up, MonoBoolean *supports_multicast)
+{
+	if (!ifname || strlen (ifname) == 0 || (!is_up && !supports_multicast))
```

Comment:
In furtherance of @jaykrell's [earlier comment about whether native code can be trusted](https://github.com/mono/mono/pull/6500/files?w=1#r189132779), this is a perfect example for why it may be hard to answer reliably: if `is_up` is NULL -- which *can't* happen when C# is invoking via `out bool is_up` -- we return early. The only way to *verify* that C# `out` is correct is to read the entire function, and on every function modification we would need to somehow "know" that this is an `out` parameter and thus `*is_up` should be assigned to when `is_up` is not NULL.

I'm inclined to agree with @jaykrell that using `ref` will be easier to mentally model over time, *especially* when there is no specific hint in this C code that `is_up` is a C# `out` parameter.